tests/tape/test_qscript.py

```diff
@@ -0,0 +1,167 @@
+        assert qs._par_info == {}
+        assert qs._trainable_params == []
+        assert qs._graph is None
+        assert qs._specs is None
+        assert qs._batch_size is None
+        assert qs._qfunc_output is None
+        assert qs.wires == qml.wires.Wires([])
+        assert qs.num_wires == 0
+        assert qs.is_sampled is False
+        assert qs.all_sampled is False
+        assert qs._obs_sharing_wires == []
+        assert qs._obs_sharing_wires_id == []
+
+    @pytest.mark.parametrize(
+        "ops",
+        (
+            [qml.S(0)],
+            (qml.S(0),),
+            (qml.S(i) for i in [0]),
+        ),
+    )
+    def test_provide_ops(self, ops):
+        """Test provided ops are coverted to lists."""
+        qs = QuantumScript(ops)
+        assert len(qs._ops) == 1
+        assert isinstance(qs._ops, list)
+        assert qml.equal(qs._ops[0], qml.S(0))
+
+    @pytest.mark.parametrize(
+        "m",
+        (
+            [qml.state()],
+            (qml.state(),),
+            (qml.state() for _ in range(1)),
+        ),
+    )
+    def test_provide_measurements(self, m):
+        """Test provided measurements are converted to lists."""
+        qs = QuantumScript(measurements=m)
+        assert len(qs._measurements) == 1
+        assert isinstance(qs._measurements, list)
+        assert qs._measurements[0].return_type is qml.measurements.State
+
+    @pytest.mark.parametrize(
+        "prep",
+        (
+            [qml.BasisState([1, 1], wires=(0, 1))],
+            (qml.BasisState([1, 1], wires=(0, 1)),),
+            (qml.BasisState([1, 1], wires=(0, 1)) for _ in range(1)),
+        ),
+    )
+    def test_provided_state_prep(self, prep):
+        """Test state prep are converted to lists"""
+        qs = QuantumScript(prep=prep)
+        assert len(qs._prep) == 1
+        assert isinstance(qs._prep, list)
+        assert qml.equal(qs._prep[0], qml.BasisState([1, 1], wires=(0, 1)))
+
+
+sample_measurements = [
+    qml.sample(),
+    qml.counts(),
+    qml.counts(all_outcomes=True),
+    qml.classical_shadow(wires=(0, 1)),
+    qml.shadow_expval(qml.PauliX(0)),
+]
+
+
+class TestUpdate:
+    """Test the methods called by _update."""
+
+    def test_cached_graph_specs_reset(self):
+
+        qs = QuantumScript()
+        qs._graph = "hello"
+        qs._specs = "something"
+
+        qs._update()
+        assert self._graph is None
+        assert self._specs is None
+
+    def test_update_circuit_info_wires(self):
+
+        prep = [qml.BasisState([1, 1], wires=(-1, -2))]
+        ops = [qml.S(0), qml.T("a"), qml.S(0)]
+        measurement = [qml.probs(wires=("a"))]
+
+        qs = QuantumScript(ops, measurement, prep)
+        assert qs.wires == qml.wires.Wires([-1, -2, 0, "a"])
+        assert qs.num_wires == 4
+
+    @pytest.mark.parametrize("sample_ms", sample_measurements)
+    def test_update_circuit_info_sampling(self, sample_ms):
+        qs = QuantumScript(measurements=[qml.state(), sample_ms])
```

Comment:
Minor: state is invalidated with other measurements on the QNode lvl.
```suggestion
        qs = QuantumScript(measurements=[qml.expval(qml.PauliZ(0)), sample_ms])
```